aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java

```diff
@@ -0,0 +1,112 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpRequestInterceptor;
+import org.apache.hc.core5.http.io.entity.StringEntity;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.iceberg.exceptions.RESTException;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.signer.Aws4Signer;
+import software.amazon.awssdk.auth.signer.internal.SignerConstant;
+import software.amazon.awssdk.auth.signer.params.Aws4SignerParams;
+import software.amazon.awssdk.auth.signer.params.SignerChecksumParams;
+import software.amazon.awssdk.core.checksums.Algorithm;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.regions.Region;
+
+public class RESTSigV4Signer implements HttpRequestInterceptor {
+  private static final String EMPTY_BODY_SHA256 =
+      "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+
+  private final Aws4Signer signer = Aws4Signer.create();
+  private AwsCredentialsProvider credentialsProvider;
+
+  private String signingName;
+  private Region signingRegion;
+
+  public void initialize(Map<String, String> properties) {
+    AwsProperties awsProperties = new AwsProperties(properties);
+
+    this.signingRegion = awsProperties.restSigv4SigningRegion();
+    this.signingName = awsProperties.restSigv4SigningName();
+    this.credentialsProvider = awsProperties.restSigv4CredentialsProvider();
+  }
+
+  @Override
+  public void process(HttpRequest request, EntityDetails entity, HttpContext context) {
+    try {
+      Aws4SignerParams params =
+          Aws4SignerParams.builder()
+              .signingName(signingName)
+              .signingRegion(signingRegion)
+              .awsCredentials(credentialsProvider.resolveCredentials())
+              .checksumParams(
+                  SignerChecksumParams.builder()
+                      .algorithm(Algorithm.SHA256)
+                      .isStreamingRequest(false)
+                      .checksumHeaderName(SignerConstant.X_AMZ_CONTENT_SHA256)
+                      .build())
+              .build();
+
+      SdkHttpFullRequest.Builder sdkRequestBuilder = SdkHttpFullRequest.builder();
+
+      sdkRequestBuilder
+          .method(SdkHttpMethod.fromValue(request.getMethod()))
+          .protocol(request.getScheme())
+          .uri(request.getUri())
```

Comment:
Only `request.getUri()` needs to be in the `try/catch`. Can you call this first and put the URI in a variable to avoid the over-broad `try` block?